src/index.ts

```diff
@@ -228,17 +228,29 @@ export function Tuple(...runtypes: Runtype<any>[]) {
 }
 
 /**
- * Construct a runtype for arbitrary dictionaries. Note that unlike Record, this provides no actual
- * runtime validation of the keys or values.
+ * Construct a runtype for arbitrary dictionaries.
  */
-export function Dictionary<V>(keyType?: 'string'): Runtype<{ [_: string]: V }>
-export function Dictionary<V>(keyType?: 'number'): Runtype<{ [_: number]: V }>
-export function Dictionary<V>() {
-  return runtype(x => {
-    if (x === null || x === undefined)
-      throw new ValidationError(`Expected a defined non-null value but was ${typeof x}`)
-
-    return x
+export function Dictionary<V>(v: Runtype<V>, keyType: 'string'): Runtype<{ [_: string]: V }>
+export function Dictionary<V>(v: Runtype<V>, keyType: 'number'): Runtype<{ [_: number]: V }>
+export function Dictionary<V>(v: Runtype<V>): Runtype<{ [_: number]: V } | { [_: string]: V }>
+export function Dictionary<V>(v: Runtype<V>, keyType?: 'string' | 'number') {
+  return runtype((xs: {[k: string]: V}) => {
+    if (xs === null || xs === undefined || typeof xs !== 'object')
+      throw new ValidationError(`Expected a defined non-null value but was ${typeof xs}`)
+    for (const key of Object.keys(xs)) {
+      // Object.keys is always a string array
+      if (keyType === 'number') {
+        if (isNaN(+key))
+          throw new ValidationError(`Expected dictionary key to be a number but was string`)
+      }
+      if (keyType === 'string') {
+        if (''+(+key) === key)
+          throw new ValidationError(`Expected dictionary key to be a string but was number`)
```

Comment:
I'm not sure this particular check is valid. This means that we disallow string dictionaries with any keys that happen to be a string representation of a number. TypeScript treats this as valid, for instance:

```ts
const foo: { [_: string]: string } = { x: 'blah', 3: 'hi' }
```

and I'd prefer to stay as close to the intended semantics of the static type system as possible.